File: app/lib/core/project_json_guard.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;

import 'path_keys.dart';
import 'project_refs.dart';

const projectJsonPathFields = {
  'source',
  'path',
  'cover_art_path',
  'backup_path',
  'package_path',
  'package_audio',
  'source_baseline_manifest',
  'source_package_manifest',
  'package_root',
  'playlist_plan',
  'timing_manifest',
  'baseline_manifest',
  'source_audio_dir',
  'source_radio_info',
  'source_bank',
  'prepared_wav',
  'staged_wav',
  'source_string_tables_dir',
  'source_table',
  'target_table',
  'packaged_table',
};

class ProjectJsonPathViolation {
  const ProjectJsonPathViolation({
    required this.pointer,
    required this.field,
    required this.value,
    required this.message,
  });

  final String pointer;
  final String field;
  final String value;
  final String message;
}

class ProjectJsonPathSchemaException implements Exception {
  const ProjectJsonPathSchemaException({
    required this.projectDir,
    required this.violations,
    this.filePath,
  });

  final String projectDir;
  final String? filePath;
  final List<ProjectJsonPathViolation> violations;

  @override
  String toString() {
    final target = filePath == null ? 'project JSON' : filePath!;
    final first = violations
        .take(5)
        .map((violation) {
          return '${violation.pointer}: ${violation.message} (${violation.value})';
        })
        .join('; ');
    final suffix = violations.length > 5
        ? '; ... ${violations.length - 5} more'
        : '';
    return 'ProjectJsonPathSchemaException: $target has '
        '${violations.length} project path schema violation(s): $first$suffix';
  }
}

List<ProjectJsonPathViolation> findProjectJsonPathViolations({
  required String projectDir,
  required Object? payload,
  Set<String> pathFields = projectJsonPathFields,
}) {
  final root = p.normalize(Directory(projectDir).absolute.path);
  final violations = <ProjectJsonPathViolation>[];

  void walk(Object? value, List<String> pointer, String? field) {
    if (value is Map) {
      for (final entry in value.entries) {
        final key = '${entry.key}';
        walk(entry.value, [...pointer, key], key);
      }
      return;
    }
    if (value is List) {
      for (var index = 0; index < value.length; index += 1) {
        walk(value[index], [...pointer, '$index'], field);
      }
      return;
    }
    if (field == null || !pathFields.contains(field) || value is! String) {
      return;
    }

    final text = value.trim();
    if (text.isEmpty) return;
    final location = _jsonPointer(pointer);
    if (isProjectRef(text)) {
      try {
        normalizeProjectRef(text);
      } on ProjectRefException catch (error) {
        violations.add(
          ProjectJsonPathViolation(
            pointer: location,
            field: field,
            value: text,
            message: error.message,
          ),
        );
      }
      return;
    }
    if (!p.isAbsolute(text)) return;
    final absolute = p.normalize(File(text).absolute.path);
    if (!isCanonicalPathInside(root, absolute)) return;
    violations.add(
      ProjectJsonPathViolation(
        pointer: location,
        field: field,
        value: text,
        message: 'project-owned absolute path must be written as fh-project:/',
      ),
    );
  }

  walk(payload, const [], null);
  return violations;
}

void assertProjectJsonPathSchema({
  required String projectDir,
  required Object? payload,
  String? filePath,
  Set<String> pathFields = projectJsonPathFields,
}) {
  final violations = findProjectJsonPathViolations(
    projectDir: projectDir,
    payload: payload,
    pathFields: pathFields,
  );
  if (violations.isEmpty) return;
  throw ProjectJsonPathSchemaException(
    projectDir: projectDir,
    filePath: filePath,
    violations: violations,
  );
}

void writeProjectJsonSync({
  required String projectDir,
  required File file,
  required Object? payload,
  JsonEncoder encoder = const JsonEncoder.withIndent('  '),
  Set<String> pathFields = projectJsonPathFields,
}) {
  assertProjectJsonPathSchema(
    projectDir: projectDir,
    payload: payload,
    filePath: file.path,
    pathFields: pathFields,
  );
  file.parent.createSync(recursive: true);
  file.writeAsStringSync(encoder.convert(payload), encoding: utf8);
}

String _jsonPointer(List<String> segments) {
  if (segments.isEmpty) return '/';
  return segments.map((segment) {
    return '/${segment.replaceAll('~', '~0').replaceAll('/', '~1')}';
  }).join();
}

File: backend/fh_radio_studio_cli/project_json_guard.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Optional

from .project_refs import (
    ProjectRefError,
    absolute_path,
    is_project_ref,
    normalize_project_ref,
    project_ref_for_path,
)

PROJECT_JSON_PATH_FIELDS = {
    "source",
    "path",
    "cover_art_path",
    "backup_path",
    "package_path",
    "package_audio",
    "source_baseline_manifest",
    "source_package_manifest",
    "package_root",
    "playlist_plan",
    "timing_manifest",
    "baseline_manifest",
    "source_audio_dir",
    "source_radio_info",
    "source_bank",
    "prepared_wav",
    "staged_wav",
    "source_string_tables_dir",
    "source_table",
    "target_table",
    "packaged_table",
}


@dataclass(frozen=True)
class ProjectJsonPathViolation:
    pointer: str
    field: str
    value: str
    message: str


class ProjectJsonPathSchemaError(ValueError):
    def __init__(
        self,
        *,
        project_dir: Path,
        violations: list[ProjectJsonPathViolation],
        path: Optional[Path] = None,
    ) -> None:
        self.project_dir = project_dir
        self.violations = violations
        self.path = path
        target = str(path) if path is not None else "project JSON"
        preview = "; ".join(
            f"{item.pointer}: {item.message} ({item.value})" for item in violations[:5]
        )
        if len(violations) > 5:
            preview += f"; ... {len(violations) - 5} more"
        super().__init__(
            f"{target} has {len(violations)} project path schema violation(s): {preview}"
        )


def find_project_json_path_violations(
    project_dir: Path,
    payload: object,
    *,
    path_fields: Iterable[str] = PROJECT_JSON_PATH_FIELDS,
) -> list[ProjectJsonPathViolation]:
    root = absolute_path(project_dir)
    fields = set(path_fields)
    violations: list[ProjectJsonPathViolation] = []

    def walk(value: object, pointer: list[str], field: Optional[str]) -> None:
        if isinstance(value, dict):
            for key, child in value.items():
                walk(child, [*pointer, str(key)], str(key))
            return
        if isinstance(value, list):
            for index, child in enumerate(value):
                walk(child, [*pointer, str(index)], field)
            return
        if field is None or field not in fields or not isinstance(value, str):
            return

        text = value.strip()
        if not text:
            return
        location = _json_pointer(pointer)
        if is_project_ref(text):
            try:
                normalize_project_ref(text)
            except ProjectRefError as exc:
                violations.append(ProjectJsonPathViolation(location, field, text, str(exc)))
            return
        path = Path(text).expanduser()
        if not path.is_absolute():
            return
        if project_ref_for_path(root, path) is None:
            return
        violations.append(
            ProjectJsonPathViolation(
                location,
                field,
                text,
                "project-owned absolute path must be written as fh-project:/",
            )
        )

    walk(payload, [], None)
    return violations


def assert_project_json_path_schema(
    project_dir: Path,
    payload: object,
    *,
    path: Optional[Path] = None,
    path_fields: Iterable[str] = PROJECT_JSON_PATH_FIELDS,
) -> None:
    violations = find_project_json_path_violations(
        project_dir,
        payload,
        path_fields=path_fields,
    )
    if not violations:
        return
    raise ProjectJsonPathSchemaError(
        project_dir=absolute_path(project_dir),
        path=path,
        violations=violations,
    )


def write_project_json(
    path: Path,
    data: object,
    *,
    project_dir: Optional[Path] = None,
    path_fields: Iterable[str] = PROJECT_JSON_PATH_FIELDS,
) -> None:
    resolved_project_dir = project_dir or project_dir_for_project_json_path(path)
    if resolved_project_dir is not None:
        assert_project_json_path_schema(
            resolved_project_dir,
            data,
            path=path,
            path_fields=path_fields,
        )
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, indent=2, ensure_ascii=False), encoding="utf-8")


def project_dir_for_project_json_path(path: Path) -> Optional[Path]:
    resolved = absolute_path(path)
    parent = resolved.parent
    if parent.name in {".fh-radio-studio", "analysis", "siren"}:
        return parent.parent
    parents = list(resolved.parents)
    if len(parents) > 3 and parents[2].name == "packages":
        return parents[3]
    if len(parents) > 2 and parents[1].name == "backups":
        return parents[2]
    if len(parents) > 3 and parents[2].name == "backups":
        return parents[3]
    return None


def _json_pointer(segments: list[str]) -> str:
    if not segments:
        return "/"
    return "".join(f"/{segment.replace('~', '~0').replace('/', '~1')}" for segment in segments)
